flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java

```diff
@@ -652,29 +587,62 @@ public void testFailingJobManagerRunnerCleanup() throws Exception {
+        final BlockingJobManagerRunnerFactory blockingJobManagerRunnerFactory =
                 new BlockingJobManagerRunnerFactory(
                         () -> {
-                            final Optional<Exception> take = queue.take();
-                            final Exception exception = take.orElse(null);
-
-                            if (exception != null) {
-                                throw exception;
+                            final Optional<Exception> maybeException = queue.take();
+                            if (maybeException.isPresent()) {
+                                throw maybeException.get();
                             }
                         });
+
+        final BlockingQueue<String> cleanUpEvents = new LinkedBlockingQueue<>();
+
+        // Track cleanup - ha-services
+        final CompletableFuture<JobID> cleanupJobData = new CompletableFuture<>();
+        haServices.setCleanupJobDataFuture(cleanupJobData);
+        cleanupJobData.thenAccept(jobId -> cleanUpEvents.add("ha-services"));
+
+        // Track cleanup - job-graph
+        final TestingJobGraphStore jobGraphStore =
+                TestingJobGraphStore.newBuilder()
+                        .setReleaseJobGraphConsumer(jobId -> cleanUpEvents.add("job-graph-release"))
+                        .setRemoveJobGraphConsumer(jobId -> cleanUpEvents.add("job-graph-remove"))
+                        .build();
+        jobGraphStore.start(null);
+        haServices.setJobGraphStore(jobGraphStore);
+
+        // Track cleanup - running jobs registry
+        haServices.setRunningJobsRegistry(
+                new StandaloneRunningJobsRegistry() {
+
+                    @Override
+                    public void clearJob(JobID jobID) {
+                        super.clearJob(jobID);
+                        cleanUpEvents.add("running-jobs-registry");
+                    }
+                });
+
         dispatcher =
                 createAndStartDispatcher(
                         heartbeatServices, haServices, blockingJobManagerRunnerFactory);
 
         final DispatcherGateway dispatcherGateway =
                 dispatcher.getSelfGateway(DispatcherGateway.class);
 
+        // submit and fail during job master runner construction
         queue.offer(Optional.of(testException));
         try {
             dispatcherGateway.submitJob(jobGraph, TIMEOUT).get();
         } catch (Throwable expectedException) {
             assertThat(expectedException, containsCause(FlinkException.class));
             assertThat(expectedException, containsMessage(testException.getMessage()));
+            // make sure we've cleaned up in correct order (including HA)
+            assertThat(
+                    new ArrayList<>(cleanUpEvents),
+                    equalTo(
+                            Arrays.asList(
+                                    "job-graph-remove", "running-jobs-registry", "ha-services")));
```

Comment:
We could introduce constants for these names. Then it is easier to maintain them.